src/parseTools.js

```diff
@@ -23,44 +23,57 @@ function processMacros(text) {
-// to locate errors for reporting.
+// to locate errors for reporting and for html files to stop expansion between <style> and </style>.
 function preprocess(text, filenameHint) {
+  var fileExt = (filenameHint) ? filenameHint.split('.').pop().toLowerCase() : "";
+  var isHtml = (fileExt === 'html' || fileExt === 'htm') ? true : false;
+  var inStyle = false;
   var lines = text.split('\n');
   var ret = '';
   var showStack = [];
   for (var i = 0; i < lines.length; i++) {
     var line = lines[i];
     try {
-      if (line[line.length-1] == '\r') {
+      if (line[line.length-1] === '\r') {
         line = line.substr(0, line.length-1); // Windows will have '\r' left over from splitting over '\r\n'
       }
-      if (!line[0] || line[0] != '#') {
-        if (showStack.indexOf(false) == -1) {
-          ret += line + '\n';
-        }
-      } else {
-        if (line[1] == 'i') {
-          if (line[2] == 'f') { // if
-            var parts = line.split(' ');
-            var after = parts.slice(1).join(' ');
-            var truthy = !!eval(after);
-            showStack.push(truthy);
-          } else if (line[2] == 'n') { // include
-            var filename = line.substr(line.indexOf(' ')+1);
-            if (filename.indexOf('"') === 0) {
-              filename = filename.substr(1, filename.length - 2);
-            }
-            var included = read(filename);
-            ret += '\n' + preprocess(included, filename) + '\n'
+      if (isHtml && line.indexOf('<style') !== -1 && !inStyle) {
+        inStyle = true;
+      }
+      if (isHtml && line.indexOf('</style') !== -1 && inStyle) {
+        inStyle = false;
+      }
+
+      if (!inStyle) {
+        if (line.indexOf('#if') === 0) {
+          var parts = line.split(' ');
+          var after = parts.slice(1).join(' ');
+          var truthy = !!eval(after);
+          showStack.push(truthy);
+        } else if (line.indexOf('#include') === 0) {
+          var filename = line.substr(line.indexOf(' ')+1);
+          if (filename.indexOf('"') === 0) {
+            filename = filename.substr(1, filename.length - 2);
           }
-        } else if (line[2] == 'l') { // else
-          assert(showStack.length > 0, 'preprocessing error parsing #else on line ' + i + ': ' + line);
+          var included = read(filename);
+          ret += '\n' + preprocess(included, filename) + '\n';
+        } else if (line.indexOf('#else') === 0) {
+          assert(showStack.length > 0);
           showStack.push(!showStack.pop());
-        } else if (line[2] == 'n') { // endif
-          assert(showStack.length > 0, 'preprocessing error parsing #endif on line ' + i + ': ' + line);
+        } else if (line.indexOf('#endif') === 0) {
+          assert(showStack.length > 0);
           showStack.pop();
         } else {
-          throw "Unclear preprocessor command on line " + i + ': ' + line;
+          if (line[0] === '#') {
+            printErr("Ignoring unclear preprocessor command: " + line);
```

Comment:
why did the throw become a warning here?

Reply:
In the discussion on the bug page, @rspq pointed out an edge case like this:

    <a href="
    #C4">Jump to Chapter 4</a>

So not throwing on unrecognised tokens seemed prudent.

Reply:
Hmm, actually that seems like a serious issue to me. If we can't error on stuff like that, maybe this isn't worth doing - if we just warn then we might miss crucial things in our JS glue setup. We've actually had bugs there in the past, that erroring helped fix.

If erroring would make us emit even worse errors potentially on some HTML, that sounds bad. In that case, I'd prefer that we not fix this, and instead find the best error we can emit for css style blocks, so that people that hit it can work around it.

Reply:
In fairness this is a pretty unlikely edge case and easily fixed if it happens - so reverting this back to an error doesn't seem unreasonable.